### PR TITLE
HPCC-25735 Speed-up std library testing.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Split STD library test into module tests like: Date, str,
uni, Math, DataPatterns with versioning.

Change TestDate.ecl to create output only when the
generated result doesn't match to the expected.

Fix parallel query execution problem in TestStore.ecl. Add
WUID to namespace identifier to avoid crossover between to parallel tests.

Exclude Date.TestDate.TestDynamic version from Thor and Roxie.

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [x] Test my draft Pull Request.

## Testing:

Tested manually.

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@ghalliday @dcamper is there any other part of the STD library test which I can restrict to one engine?

**Comment by AttilaVamos** [comment]
It is still WIP that's why there are two stdlibrarytest ECl files. In the final version there would be only one.

**Review by dcamper — APPROVED** [review]

**Comment by dcamper** [comment]
> @ghalliday @dcamper is there any other part of the STD library test which I can restrict to one engine?

I would think that the tests Date/*.ecl, Math/*.ecl, Metaphone/*.ecl, str/*.ecl and uni/*.ecl could all be done on a single engine because they all either exercise calls in rtl or system calls, both of which are common.  Gavin will probably tell me I'm wrong, though.

system/TestStore.ecl is really exercising SOAPCALL.  That could be constrained to either hthor or ROXIE but not necessarily both.  Including thor may be a good idea because SOAPCALL is handled specially there.

**Comment by AttilaVamos** [comment]
@dcamper @ghalliday @jakesmith 

Runtime information of STD library modules in On-Demand/AWS based and GH Action based Smoketest (in last test result) :

| Module | AWS Hthor | AWS Thor | AWS Roxie | - | GH Hthor | GH Thor | GH Roxie |
| ---- | ---- | ---- | ---- | ----| ---- | ---- | ---- |
| BLAS | 2 s | 4 s | 2 s | | 1 s |2 s|2 s |
|Crypto|2 s|5 s|3 s| |1 s|2 s |2 s |
|DataPatterns|82 s|229 s|40 s| |120 s|236 s |211 s |
|TestDate.TestDynamic|4 s| excl. | excl. | | 3 s| excl. | excl. |
|TestFormat|2 s|5 s|3 s| |1 s|2 s |2 s |
|File|2 s|8 s|3 s| |2 s|2 s |1 s |
|Math|2 s|4 s|14 s| |1 s|1 s |2 s |
|Metaphone|2 s|6 s|2 s | |1 s|2 s |1 s |
|str|3 s|4 s|16 s| | 1 s|2 s |2 s |
|system|4 s|239 s|4 s| |3 s|5 s |4 s |
|uni|2 s|7 s|3 s| |2 s|3 s |2 s |

Original teststdlibrary.ecl times in last 60 test session on AWS based Smoketest
| Engine | Min (sec) | Max (sec) | Average (sec)|
| ----| ----| ----| ----|
|Hthor|59|115|90|
|Thor|234|663|272|
|Roxie|38|78|55|

File: ecllibrary/teststd/Date/TestDate.ecl [hunk 1]
```diff
@@ -435,8 +435,7 @@ EXPORT TestDate := MODULE
     END;
 
     x := LOOP(firstDate, 1000000, processNextDate(ROWS(LEFT)));
-
-    EXPORT Test01 := OUTPUT(x);
+    EXPORT Test01 := ASSERT( x[1].day = 1000001 AND x[1].gregorian.year = 2738 AND x[1].gregorian.month = 11 AND x[1].gregorian.day = 29 AND x[1].julian.year = 2738 AND x[1].julian.month = 11 AND x[1].julian.day = 8, 'Failed') ;
```
Comment (ghalliday) [inline ecllibrary/teststd/Date/TestDate.ecl:438]: minor: Would benefit of being split over multiple lines.

File: testing/regress/ecl/teststdlibrary2.ecl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import ^ as root;
+
+unsigned VERBOSE := 0;
+
+string myMod := #IFDEFINED(root.tmod, 'Full');
+#if (VERBOSE = 1)
+    output(myMod, named('myMod'));
+#end
+
+#IF (myMod = 'Date.TestDate.TestDynamic' )
```
Comment (ghalliday) [inline testing/regress/ecl/teststdlibrary2.ecl:45]: I think you could probably replace most of this nest of #ifs with

`import #expand('teststd.' + myMod) as testMod;
`

**Review by ghalliday — COMMENTED** [review]
A couple of fairly minor comments.  General approach looks good.

File: ecllibrary/teststd/Date/TestDate.ecl (hunk 1 above)
Comment (AttilaVamos) [reply ecllibrary/teststd/Date/TestDate.ecl]: Something like this?

File: testing/regress/ecl/teststdlibrary2.ecl (hunk 2 above)
Comment (AttilaVamos) [reply testing/regress/ecl/teststdlibrary2.ecl]: Yes it is a much smarter solution. Changed.

**Review by ghalliday — APPROVED** [review]
@AttilaVamos commit message still contains WIP, otherwise it looks good to merge.

**Comment by ghalliday** [comment]
And needs to become a non-draft PR

**Comment by AttilaVamos** [comment]
@ghalliday please check and merge.

**Pull request merged**